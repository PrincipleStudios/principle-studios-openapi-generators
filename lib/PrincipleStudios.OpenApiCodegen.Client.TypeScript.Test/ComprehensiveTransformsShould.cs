using PrincipleStudios.OpenApi.Transformations;
using PrincipleStudios.OpenApi.TypeScript;
using Snapshooter.Xunit;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using static PrincipleStudios.OpenApiCodegen.TestUtils.DocumentHelpers;

namespace PrincipleStudios.OpenApiCodegen.Client.TypeScript
{
    using static OptionsHelpers;
    public class ComprehensiveTransformsShould
    {
        [Trait("Category", "Integration")]
        [MemberData(nameof(ValidFileNames))]
        [Theory]
        public void CoverFullFiles(string name)
        {
            var document = GetDocument(name);
            var options = LoadOptions();

            var transformer = document.BuildTypeScriptOperationSourceProvider("", options);
            OpenApiTransformDiagnostic diagnostic = new();

            var entries = transformer.GetSources(diagnostic).ToArray();

            var sb = new StringBuilder();
            foreach (var entry in entries)
            {
                sb.AppendLine($"exports[`{entry.Key}`] = `{entry.SourceText}`;");
                sb.AppendLine();
            }
            Snapshot.Match(sb.ToString(), $"{nameof(ComprehensiveTransformsShould)}.{TypeScriptNaming.ToTitleCaseIdentifier(name, options.ReservedIdentifiers())}");
            Assert.Empty(diagnostic.Errors);
        }

        [Trait("Category", "Integration")]
        [MemberData(nameof(InvalidFileNames))]
        [Theory]
        public void ReportDiagnosticsForMissingReferences(string name)
        {
            var document = GetDocument(name);
            var options = LoadOptions();

            var transformer = document.BuildTypeScriptOperationSourceProvider("", options);
            OpenApiTransformDiagnostic diagnostic = new();

            var entries = transformer.GetSources(diagnostic).ToArray();

            Snapshot.Match(diagnostic.Errors.Select(err => new { Context = err.Context.ToOpenApiPathContextString(), Message = err.Message }).ToArray(), $"Diagnostics.{TypeScriptNaming.ToTitleCaseIdentifier(name, options.ReservedIdentifiers())}");
        }

        public static IEnumerable<object[]> ValidFileNames =>
            from fileIndex in GetValidDocumentIndices()
            select new object[] { GetDocumentName(fileIndex) };

        public static IEnumerable<object[]> InvalidFileNames =>
            from fileIndex in GetInvalidDocumentIndices()
            select new object[] { GetDocumentName(fileIndex) };

    }
}
