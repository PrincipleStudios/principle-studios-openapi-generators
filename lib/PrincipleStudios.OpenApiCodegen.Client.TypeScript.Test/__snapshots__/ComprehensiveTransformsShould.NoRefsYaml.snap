exports[`operations/lookupRecord.ts`] = `/*
 * No Refs Demo
 *
 * A sample API that uses no refs
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { HttpHeaders, HttpQuery, throwIfNullOrUndefined, encodeURI, RequestConversions, RequestOpts, ResponseArgs, StandardResponse } from '@principlestudios/openapi-codegen-typescript';

export type RequestParams = {
};
export type RequestBodies = {
    'application/json': {
        body: { "formattedAddress": string; "location": { "latitude": number; "longitude": number } };
    };
};

export type Responses = {
    /** unexpected error */
    'other': {
        'application/json': { "code": string; "message": string };
    };
    200: {
        'application/json': { "id": string };
    };
    409: {
        'application/json': { "multiple": Array<{ "formattedAddress": string; "location": { "latitude": number; "longitude": number } }> };
    };
};

export function lookupRecord<K extends keyof RequestBodies>(params: RequestParams, body: RequestBodies[K], mimeType: K): RequestOpts;
export function lookupRecord(params: RequestParams, body?: RequestBodies[keyof RequestBodies], mimeType?: keyof RequestBodies): RequestOpts {
    throwIfNullOrUndefined(params, 'params');
    if (mimeType) throwIfNullOrUndefined(body, 'body');
    if (body) throwIfNullOrUndefined(mimeType, 'mimeType');

    const headers: HttpHeaders = {
        'Content-Type': mimeType,
    };

    return {
        path: '/address',
        method: 'POST',
        headers,
        body,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            200,
            409,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.xhr.getResponseHeader("Content-Type"),
        data: res.response,
        response: res,
    }) as StandardResponse<Responses>;

export const conversion: RequestConversions<RequestParams, | RequestBodies, Responses> = {
    request: lookupRecord,
    response: responseConversion,
};

`;

exports[`operations/getPhoto.ts`] = `/*
 * No Refs Demo
 *
 * A sample API that uses no refs
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { HttpHeaders, HttpQuery, throwIfNullOrUndefined, encodeURI, RequestConversions, RequestOpts, ResponseArgs, StandardResponse } from '@principlestudios/openapi-codegen-typescript';

export type RequestParams = {
    id: string;
};
export type RequestBodies = {
};

export type Responses = {
    /** unexpected error */
    'other': {
        'application/json': { "code": string; "message": string };
    };
    200: {
        'application/octet-stream': string;
    };
};

export function getPhoto(params: RequestParams): RequestOpts;
export function getPhoto(params: RequestParams): RequestOpts {
    throwIfNullOrUndefined(params, 'params');
    throwIfNullOrUndefined(params.id, 'id');

    const headers: HttpHeaders = {
    };

    return {
        path: '/photo/{id}'.replace('{id}', encodeURI(params.id)),
        method: 'GET',
        headers,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            200,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.xhr.getResponseHeader("Content-Type"),
        data: res.response,
        response: res,
    }) as StandardResponse<Responses>;

export const conversion: RequestConversions<RequestParams,  | undefined, Responses> = {
    request: getPhoto,
    response: responseConversion,
};

`;

exports[`operations/index.ts`] = `/*
 * No Refs Demo
 *
 * A sample API that uses no refs
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { conversion as lookupRecord } from './lookupRecord';
import { conversion as getPhoto } from './getPhoto';

export default {
    lookupRecord,
    getPhoto,
} as const;
`;

